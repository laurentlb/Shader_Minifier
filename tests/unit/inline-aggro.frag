// *** Things we should inline

float inl1() {
	float f = 42.0;
	float g = 2.0 * f;
	return f + g;
}

float inl2(float x) {
	float f = 123456789.0;
	if (x > 2.0*f) return 100.0;
	if (x > f) return 101.0;
	return 102.0;
}

float inl3() {
	const float f = acos(-1.0);
	const float g = 2.0*f;
	return 0.75*g;
}

float inl4() {
	const float f = acos(-1.0), g = 2.0*f;
	return 0.75*g;
}

const float foo = 123.0;
float inl5() {
	float f = 456.0;
	return foo + f;
}

const float bar = acos(-1.0);
const float baz = 2.0*bar;
float inl6() {
	return 2.0 * baz;
}

void notevil(in float x) { x = 42.0; }
float inl7() {
	float f = 101.0;
	notevil(f);
	return f;
}

int inl8(in ivec3 x) {
	int i = 1;

	// i is not technically in an lvalue position but not smart enough right now to see that.
	x[i] += 1;
	return x[i] + i;
}

// *** Things we shouldn't inline

float noinl1() {
	float f = 42.0;
	f = 101.0;
	return f;
}

float noinl2() {
	float f = 42.0;
	f++;
	return f;
}

float noinl3() {
	float f = 42.0;
	if (acos(-1.0) < 3.0) {
		f = 101.0;
	}
	return f;
}

float noinl4() {
	float f = 42.0;
	for (f = 0.0; false;);
	return f;
}

float noinl5() {
	float f = 42.0;
	for (; false; f++);
	return f;
}

float noinl6(float x) {
	float f = x + 1.0;
	x = 100.0;
	return f + 2.0;
}

float noinl7(const float x) {
	return x + 1.2;
}

float quux = 1.0;
float noinl8() {
	return quux + 2.0;
}

float noinl9(float x) {
	float bar = x + 1.0; // Shadows inlinable const above
	x = 100.0;
	return bar + 2.0;
}

void evil(inout float x) { x = 42.0; }
float noinl10() {
	float f = 101.0;
	evil(f);
	return f;
}

int noinl11(in ivec3 x) {
	int i = 1;
	x[i++] += 1;
	return x[i] + i;
}

float noinl12() {
	int i = 10;
	while (--i > 0) {
	}
	return 1;
}
