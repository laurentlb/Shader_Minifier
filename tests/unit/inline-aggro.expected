float inl1()
{
  return 126.;
}
float inl2(float x)
{
  return x>246913578.?
    1e2:
    x>123456789.?
      101.:
      102.;
}
float inl3()
{
  return 2.*acos(-1.)*.75;
}
float inl4()
{
  return 2.*acos(-1.)*.75;
}
float inl5()
{
  return 579.;
}
const float bar=acos(-1.),baz=2.*bar;
float inl6()
{
  return 2.*baz;
}
void notevil()
{
  float x=42.;
}
float inl7()
{
  notevil();
  return 101.;
}
int inl8(ivec3 x)
{
  x[1]+=1;
  return x[1]+1;
}
float noinl1()
{
  return 101.;
}
float noinl2()
{
  float f=42.;
  f++;
  return f;
}
float noinl3()
{
  float f=42.;
  if(acos(-1.)<3.)
    f=101.;
  return f;
}
float noinl4()
{
  float f=42.;
  for(f=0.;false;)
    ;
  return f;
}
float noinl5()
{
  float f=42.;
  for(;false;f++)
    ;
  return f;
}
float noinl6(float x)
{
  float f=x+1.;
  x=1e2;
  return f+2.;
}
float noinl7(const float x)
{
  return x+1.2;
}
float noinl8()
{
  return 3.;
}
float noinl9(float x)
{
  float bar=x+1.;
  x=1e2;
  return bar+2.;
}
void evil(inout float x)
{
  x=42.;
}
float noinl10()
{
  float f=101.;
  evil(f);
  return f;
}
int noinl11(ivec3 x)
{
  int i=1;
  x[i++]+=1;
  return x[i]+i;
}
int noinl12()
{
  int i=10;
  for(;--i>0;)
    ;
  return 1;
}
